x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -43,746 +40,588 @@
-                if (Strings.hasText(clientAuthorizationSharedSecret)) {
-                    throw new SettingsException(
-                        "Setting ["
-                            + RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.CLIENT_AUTHORIZATION_SHARED_SECRET)
-                            + "] is not supported, because setting ["
-                            + RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.CLIENT_AUTHORIZATION_TYPE)
-                            + "] is ["
-                            + JwtRealmSettings.SUPPORTED_CLIENT_AUTHORIZATION_TYPE_NONE
-                            + "]"
-                    );
-                }
-                break;
+    private Cache<String, JwtRealmCacheValue> initializeJwtCache() {
+        final TimeValue cacheTtl = super.config.getSetting(JwtRealmSettings.CACHE_TTL);
+        if (cacheTtl.getNanos() > 0) {
+            final Integer cacheMaxUsers = super.config.getSetting(JwtRealmSettings.CACHE_MAX_USERS);
+            return CacheBuilder.<String, JwtRealmCacheValue>builder().setExpireAfterWrite(cacheTtl).setMaximumWeight(cacheMaxUsers).build();
         }
+        return null;
     }
 
-    // This logic is in a method for easy unit testing
-    public static void validateIssuerCredentialSettings(
-        final RealmConfig realmConfig,
-        final SecureString hmacSecretKey,
-        final String jwkSetPath,
-        final List<String> allowedSignatureAlgorithms
-    ) throws SettingsException {
-        final boolean hasHmacSecretKey = Strings.hasText(hmacSecretKey);
-        final boolean hasJwkSetPath = Strings.hasText(jwkSetPath);
-
-        // Validate HMAC or JWK Set are set, or both.
-        if ((hasHmacSecretKey == false) && (hasJwkSetPath == false)) {
-            throw new SettingsException(
-                "At least one setting is required for ["
-                    + RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.ISSUER_HMAC_SECRET_KEY)
-                    + "] or ["
-                    + RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.JWKSET_PATH)
-                    + "]"
+    // must call parseAlgsAndJwksHmac() before parseAlgsAndJwksPkc()
+    private Tuple<List<String>, List<JWK>> parseAlgsAndJwksHmac() {
+        final SecureString jwkSetContentsHmac = super.config.getSetting(JwtRealmSettings.JWKSET_HMAC_CONTENTS);
+        if (Strings.hasText(jwkSetContentsHmac) == false) {
+            return new Tuple<>(Collections.emptyList(), Collections.emptyList());
+        }
+        List<JWK> jwksHmac; // Parse as JWKSet, or fall back to byte array
+        try {
+            jwksHmac = JwkValidateUtil.loadJwksFromJwkSetString(
+                RealmSettings.getFullSettingKey(super.config, JwtRealmSettings.JWKSET_HMAC_CONTENTS),
+                Strings.hasText(jwkSetContentsHmac) ? jwkSetContentsHmac.toString() : null
             );
+        } catch (Exception e) {
```

Comment:
I agree. Spliting HMAC JWKSet vs key is better.

I am thinking PKC JWKSet URL vs file can be split too. It will clean up some corner cases in that code.